builder-manifest: Port to libappstream-compose API

bbhtt opened the pull request:
Subject to API stability

1.5.0 unstable material. 

~This does not compile, as the asc-fonts private header is missing from appstream installation. The appstream-compose header needs to be patched right now to get it building~ Solved by stub header.

bbhtt commented on src/builder-manifest.c:
```diff
@@ -2404,26 +2407,63 @@ cmpstringp (const void *p1, const void *p2)
+
+  // TODO: Add CLI arg for these two
+  asc_compose_remove_flags (compose, ASC_COMPOSE_FLAG_ALLOW_SCREENCASTS);
+  asc_compose_add_flags (compose, ASC_COMPOSE_FLAG_NO_PARTIAL_URLS);
```
it's a bit expensive to store and serve over ostree refs and no app store i know supports it. the plan was to disable it by default and have a cli arg to enable it for those who want.

bbhtt replied:
looks like at least partial urls needs newer appstream than 0.15 something

ximion replied:
If you don't provide the data, app stores won't support it ;-) - It was a highly requested feature for ages.

> looks like at least partial urls needs newer appstream than 0.15 something

Version 0.16.3 from 2 years ago.

bbhtt commented on src/builder-manifest.c:
```diff
@@ -2404,26 +2407,63 @@ cmpstringp (const void *p1, const void *p2)
+
+  if (asc_compose_has_errors (compose))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
```
added the hints dir, i looked at the composecli_print function and it looked complicated..

bbhtt commented:
Which version fixed this?

```
 In file included from /usr/include/appstream-compose/appstream-compose.h:32,
                 from ../src/builder-manifest.c:34:
/usr/include/appstream-compose/asc-globals.h:65:1: error: function declaration isn’t a prototype [-Werror=strict-prototypes]
   65 | gchar                           **asc_globals_get_hint_tags ();
      | ^~~~~
cc1: some warnings being treated as errors
```

ximion commented:
Ubuntu wouldn't update to new major releases of basically anything anway, AFAIK it never updated an AppStream version after release.

The new release is rather trivial to backport though, libfyaml is available everywhere, also on Ubuntu 24.04.

bbhtt committed:
builder-manifest: Disable processing screencasts in appstream

Screencasts are generally unwanted in catalogue to as they are more
expensive to download, store in ref and serve. Also no appstore
currently supports showing them. This can later be changed if needed.

bbhtt committed:
builder-manifest: Allow all custom appstream tags in catalogue

This allows all `<custom>` tags from metainfo files to reach
catalogue data. Custom tags are used by Flathub to store verification
and build information and often downstream vendors who use Flathub's
catalogue for their own app store want custom tags to identify or
highlight their own apps. For example cosmic using Flathub catalogue
data to highlight cosmic apps on cosmic store app etc.

Allowing individual custom tags one by one is annoying and this is what
libappstream-glib did before the port in 1.3.4. So this brings back
the old as-glib behaviour that was lost in >=1.3.4

bbhtt committed:
builder-manifest: Set hint dir for appstream compose

On request from maintainer. This sets a place to post output
of compose errors

bbhtt committed:
builder-manifest: Port to libappstream-compose API

bbhtt committed:
builder-manifest: Port to libappstream-compose API

This also simplifies the compose handling by moving all related path
and context logic into builder_appstreamcli_compose() as this is only
ever used for Flatpak and the paths stay constant for all input
build refs.

This also allows us to remove the duplicate call due to media_dir and
mirror url being set to NULL in one branch.

The rest is an exact 1:1 port of what was present before from our side.

bbhtt committed:
[drop] Add a stub header for asc-font

bbhtt committed:
[drop] Define stuff

bbhtt commented on src/builder-manifest.c:
```diff
@@ -30,6 +30,25 @@
+#pragma GCC diagnostic ignored "-Wstrict-prototypes"
+#endif
+
+#define I_KNOW_THE_APPSTREAM_COMPOSE_API_IS_SUBJECT_TO_CHANGE
```
Looks like it is dropped in main, but no release yet. I thought I saw it dropped in 1.1.2 so I was eagerly trying to update to the CI containers. 

I will wait for a release then just try to merge this in with this define as is. I know Ubuntu will not update it in time, and I don't want to build stuff just for CI.

bbhtt replied:
@ximion any idea when can we expect 1.1.3?

I'd like to have this PR land and get 1.5.0 unstable out in the next month or so.

ximion replied:
1.1.3 probably quite soon, but I will likely revert the API stability note change for that release. The release after 1.1.3 will contain some refactoring to sandbox all media processing that the generator does (due to some incidents), and that will result in some API changes.

However, I think the code in this PR is pretty safe, as you aren't using `AscImage` directly, and all changes will happen there.

bbhtt replied:
Ah, I will wait in that case.

On incidents I was wondering it would be nice to pass fds to the appstream api instead of paths. There aren't risks but I wanted to harden the a bit and make sure everything is anchored to app_root. But I don't think I can because I can validate that and anchor them here using fd but as soon as I get a path out of that and pass it to appstream api, there are risks of tocttou.

bbhtt committed:
builder-manifest: Allow all custom appstream tags in catalogue

This allows all `<custom>` tags from metainfo files to reach
catalogue data. Custom tags are used by Flathub to store verification
and build information and often downstream vendors who use Flathub's
catalogue for their own app store want custom tags to identify or
highlight their own apps. For example cosmic using Flathub catalogue
data to highlight cosmic apps on cosmic store app etc.

Allowing individual custom tags one by one is annoying and this is what
libappstream-glib did before the port in 1.3.4. So this brings back
the old as-glib behaviour that was lost in >=1.3.4

bbhtt committed:
builder-manifest: Disable processing screencasts in appstream

Screencasts are generally unwanted in catalogue to as they are more
expensive to download, store in ref and serve. Also no appstore
currently supports showing them to my knowledge. This can later be
changed by adding an argument if needed.

bbhtt committed:
builder-main: Flip the default BuilderAsUrlPolicy to full

See: https://github.com/flatpak/flatpak-builder/issues/743

bbhtt committed:
tests: Add asan suppressions for appstream

bbhtt committed:
Revert "tests: Add asan suppressions for appstream"

This reverts commit bc36aa1725f8c202d9670d2553af0189418035c2.

bbhtt committed:
test

bbhtt committed:
add asan

ximion commented:
I intend to freeze the API with the upcoming AppStream 1.2.0 release (likely to be released this month). Looking at this patch, there don't seem to be any breaking changes that impact you, however, since AppStream 1.2.0 defaults to JPEG-XL for images, you may want to call `asc_compose_set_image_format (compose, ASC_IMAGE_FORMAT_PNG);` to restore the exact previous output.

Future versions of compose may also fully sandbox media processing, but once that is implemented, you will also be able to opt-out in the API in case that causes any issues (you can use the `AS_CHECK_VERSION` macro to test for the version of appstream(-compose) that you are building against, for compile-time switches).

Please let me know if the current API works for you. I will do some extensive testing as well before releasing it.

ximion commented:
AppStream 1.2.0 is out now (blogpost pending, this release has a lot!). Don't expect any incompatible changes for the compose API in the near future (but tbh, even this one had very few if you weren't using complex features like appstream-generator does).

AS 1.2.0 won't be backportable to Ubuntu 24.04, but once it received enough real-world testing and if it helps you in any way, I could backport it to the Ubuntu 26.04 Flatpak PPA.

bbhtt commented:
I think backporting to the PPA is a good idea. I'll end up using whatever distro will have it in CI.

ximion commented:
So, Debian Testing will have it very soon (pending review), and Arch Linux already has it. Fedora is somehow very outdated... Ubuntu 26.10 will also ship with it, and I might update the PPA for 26.04 if it helps you (but that will need a bit of testing first, users apparently use it and I don't want to break any other software by accident (even though that chance is extremely low)).
